### PR TITLE
Update ESPnet2 JSUT TTS recipe and TTS template

### PR DESCRIPTION
**Pull request by kan-bayashi**
This PR updates JSUT recipe to be almost the same as ESPnet1 recipe.
Now I'm training both `feats_type=fbank` and `feats_type=raw` models.

## Modification of TTS template

- Changed `tts_stats_dir` naming [`c8c2d16` (#2110)](https://github.com/espnet/espnet/pull/2110/commits/c8c2d1608bcf845d4281512ecd40a4ba98decda0)
- Changed default decoding model to `train.loss.best.pth` since the validation best model is not always the best in terms of the quality. I'm wondering to use `latest.pth`.
- ~~Keep consistency none or null [`6601ccf` (#2110)](https://github.com/espnet/espnet/pull/2110/commits/6601ccf48ed5f727b21a0ac2e4bde30250527540)~~ -> error happened, reverted.
- Use `n_shift` value to decide `_fold_length` for `feats_type=raw` [`5c45f62` (#2110)](https://github.com/espnet/espnet/pull/2110/commits/5c45f623077be890911dc6dbe3a3ba90848caaee)
- Support attention plot in inference
- Support stop prediction plot in inference
- Fix `normalize.inverse` in-place operation issue
- Change to use npy format to dump the generated features

### CODE REVIEW

**Comment by kan-bayashi** [comment]
@kamo-naoyuki I found that the number of minibatch is different when `batch_type: numel`.

## feats_type = fbank

```
[huracan] 2020-06-27 12:16:14,163 (abs_task:1400) INFO: [train] dataset:
ESPnetDataset(
  text: {"path": "dump/fbank/tr_no_dev/text", "type": "text"}
  speech: {"path": "dump/fbank/tr_no_dev/feats.scp", "type": "kaldi_ark"}
  preprocess: <espnet2.train.preprocessor.CommonPreprocessor object at 0x7fabc1376e48>)
[huracan] 2020-06-27 12:16:14,163 (abs_task:1401) INFO: [train] Batch sampler: NumElementsBatchSampler(N-batch=252, batch_bins=1000000, sort_in_batch=descending, sort_batch=descending)
[huracan] 2020-06-27 12:16:14,163 (abs_task:1403) INFO: [train] mini-batch sizes summary: N-batch=252, mean=28.6, min=3, max=60
[huracan] 2020-06-27 12:16:14,210 (abs_task:1400) INFO: [valid] dataset:
ESPnetDataset(
  text: {"path": "dump/fbank/dev/text", "type": "text"}
  speech: {"path": "dump/fbank/dev/feats.scp", "type": "kaldi_ark"}
  preprocess: <espnet2.train.preprocessor.CommonPreprocessor object at 0x7fabc13765c0>)
[huracan] 2020-06-27 12:16:14,210 (abs_task:1401) INFO: [valid] Batch sampler: NumElementsBatchSampler(N-batch=7, batch_bins=1000000, sort_in_batch=descending, sort_batch=descending)
[huracan] 2020-06-27 12:16:14,210 (abs_task:1403) INFO: [valid] mini-batch sizes summary: N-batch=7, mean=35.7, min=15, max=52
```

## feats_type = raw

```
[huracan] 2020-06-27 12:21:04,833 (abs_task:1400) INFO: [train] dataset:
ESPnetDataset(
  text: {"path": "dump/raw/tr_no_dev/text", "type": "text"}
  speech: {"path": "dump/raw/tr_no_dev/wav.scp", "type": "sound"}
  preprocess: <espnet2.train.preprocessor.CommonPreprocessor object at 0x7fa406350160>)
[huracan] 2020-06-27 12:21:04,833 (abs_task:1401) INFO: [train] Batch sampler: NumElementsBatchSampler(N-batch=783, batch_bins=1000000, sort_in_batch=descending, sort_batch=descending)
[huracan] 2020-06-27 12:21:04,834 (abs_task:1403) INFO: [train] mini-batch sizes summary: N-batch=783, mean=9.2, min=1, max=23
[huracan] 2020-06-27 12:21:04,874 (abs_task:1400) INFO: [valid] dataset:
ESPnetDataset(
  text: {"path": "dump/raw/dev/text", "type": "text"}
  speech: {"path": "dump/raw/dev/wav.scp", "type": "sound"}
  preprocess: <espnet2.train.preprocessor.CommonPreprocessor object at 0x7fa406350940>)
[huracan] 2020-06-27 12:21:04,875 (abs_task:1401) INFO: [valid] Batch sampler: NumElementsBatchSampler(N-batch=22, batch_bins=1000000, sort_in_batch=descending, sort_batch=descending)
[huracan] 2020-06-27 12:21:04,875 (abs_task:1403) INFO: [valid] mini-batch sizes summary: N-batch=22, mean=11.4, min=2, max=17
```

Is it intentional?

**Comment by kamo-naoyuki** [comment]
When feats_type=raw, the input dimension is the number of samples in the wav file, while frame_len * mel_freqs when feats_type=fbank.

**Comment by kan-bayashi** [comment]
I see! To match the condition, I will use `batch_type=folded` now.

**Comment by kan-bayashi** [comment]
@kamo-naoyuki I made some modifications to the template (Summarized in https://github.com/espnet/espnet/pull/2110#issue-440854190). Please check them.

**Comment by kamo-naoyuki** [comment]
> * Changed `tts_stats_dir` naming [`c8c2d16` (#2110)]

Thank you, I was also going to change it. Could you change it for asr recipe?

**Comment by kan-bayashi** [comment]
Sure. For quick fix, I made another PR #2111.

**Comment by kan-bayashi** [comment]
@kamo-naoyuki I figure out the reason of the failure of stop prediction.
It seems that <eos> symbol is not added at the end of the input sequence.

**Comment by kan-bayashi** [comment]
In the TTS case, `<eos>` is added in data loader part in espnet1.
So we need to add eos concatenation in the `forward` and `inference`.

**Comment by kan-bayashi** [comment]
Oh, `forward` already have the concatenation.
https://github.com/espnet/espnet/blob/cb4e446ee220d3a52fd5b05dd85a283424163515/espnet2/tts/tacotron2.py#L266-L270
Then, we only need to fix `inference`.

**Comment by kamo-naoyuki** [comment]
Oh, thank you! I was careless on that.

**Comment by kan-bayashi** [comment]
Not stopped yet...
Maybe this part is wrong.
https://github.com/espnet/espnet/blob/cb4e446ee220d3a52fd5b05dd85a283424163515/espnet2/tts/tacotron2.py#L276
I will check it.

**Comment by kan-bayashi** [comment]
It should be:
```python
        # make labels for stop prediction
        labels = make_pad_mask(olens - 1).to(ys.device, ys.dtype)
        labels = F.pad(labels, [0, 1], "constant", 1.0)
```

**Comment by kamo-naoyuki** [comment]
Sorry, I don't know the desired shape. Seems to make no changing?

```python
In [16]: olens = torch.tensor([3,4,5])                                                                                                 
In [17]: labels = make_pad_mask(olens - 1)                                                                                             
In [18]: labels                                                                                                                        
Out[18]: 
tensor([[0, 0, 1, 1],
        [0, 0, 0, 1],
        [0, 0, 0, 0]], dtype=torch.uint8)
In [19]: labels = F.pad(labels, [0, 1], "constant", 1.0)                                                                               
In [20]: labels                                                                                                                        
Out[20]: 
tensor([[0, 0, 1, 1, 1],
        [0, 0, 0, 1, 1],
        [0, 0, 0, 0, 1]], dtype=torch.uint8)
```

**Comment by kan-bayashi** [comment]
> Sorry, I don't know the desired shape. Seems to make no changing?

The shape is the same but the last value is different.
`labels` is the ground truth of the stop token prediction (binary classification).
Therefore, the value of the last frame must be 1.

**Comment by kan-bayashi** [comment]
Now correctly stopped with epoch 15 model :)

**Comment by kamo-naoyuki** [comment]
Thank you!

**Comment by kan-bayashi** [comment]
The modification is finished.
I will wait the training.
Once finished, I will upload the results.

File: egs2/jsut/tts1/conf/tuning/train_tacotron2.yaml [hunk 1]
```diff
@@ -52,21 +53,25 @@ optim_conf:           # keyword arguments for selected optimizer
...
+batch_bins: 1000000         # batch bins (for feats_type=fbank)
+# batch_bins: 3750000       # batch bins (for feats_type=raw, *= n_shift / n_mels)
 batch_type: numel           # how to make batch
 sort_in_batch: descending   # how to sort data in making batch
 sort_batch: descending      # how to sort created batches
 num_workers: 1              # number of workers of data loader
 train_dtype: float32        # dtype in training
 log_interval: null          # log interval in iterations
-keep_nbest_models: 1        # number of models to keep
+keep_nbest_models: 5        # number of models to keep
```
Comment (kamo-naoyuki) [inline egs2/jsut/tts1/conf/tuning/train_tacotron2.yaml:68]: Do you intend to use averaged model?
Which is better averaged model and best model? Just a question.

Comment (kan-bayashi) [reply egs2/jsut/tts1/conf/tuning/train_tacotron2.yaml:68]: Yes. Actually, I have not yet compared in detail, but I want to check it.

**Comment by kan-bayashi** [comment]
I found that the scale of mel-spectrogram is strange (generated mel-spectrogam does not keep 0 mean).
I will investigate it.

**Comment by kan-bayashi** [comment]
I figure out the reason.
`normalize.inverse` operation seems in-place.
So `outs` is replaced with de-normalized values.
https://github.com/espnet/espnet/blob/cb4e446ee220d3a52fd5b05dd85a283424163515/espnet2/bin/tts_inference.py#L131-L138

**Comment by kan-bayashi** [comment]
Now I canfirmed that the generated mel-spectrogram can be used as the input of my pretrained vocoders!

**Comment by kan-bayashi** [comment]
Maybe we want to unify the hyperparameter naming as the same as ASR models.
So now I will not upload the pretrained model.
